File: web/activity/src/player/masterPlaylist.ts
import { environment } from '../environment';
import type { Manifest } from '../types';

/**
 * Ingest writes one media playlist per rendition and no master, so the player composes one.
 *
 * The master is what binds the audio renditions to the video: without it hls.js loads a
 * video-only playlist and the film plays silently with no track to switch to. It is handed over
 * as a blob, which has no base to resolve against, so every URI in it is absolute.
 *
 * CODECS is deliberately absent. hls.js reads the real codecs out of the fMP4 initialisation
 * segments, and a declared string that disagrees with them costs a source buffer.
 */
const AudioGroupId = 'audio';

export interface MasterPlaylist {
  playlist: string;
  /** The rendition NAME written for each manifest audio track id, which is how a chosen track is found again. */
  audioNames: Record<string, string>;
}

function quoted(value: string): string {
  return value.replace(/"/g, "'");
}

export function buildMasterPlaylist(manifest: Manifest): MasterPlaylist {
  const media = (relative: string) => environment().mediaUrl(manifest.id, relative);

  const defaultAudio = manifest.audio.find((track) => track.default) ?? manifest.audio[0];
  const audioNames: Record<string, string> = {};
  const taken = new Set<string>();

  const lines = ['#EXTM3U', '#EXT-X-VERSION:7', '#EXT-X-INDEPENDENT-SEGMENTS'];

  for (const track of manifest.audio) {
    // NAME identifies a rendition inside its group, so a repeated label needs breaking apart.
    let name = quoted(track.label);
    while (taken.has(name)) name = `${name} (${track.id})`;
    taken.add(name);
    audioNames[track.id] = name;

    const isDefault = track === defaultAudio;
    lines.push(
      `#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="${AudioGroupId}",NAME="${name}",` +
        `LANGUAGE="${quoted(track.language)}",` +
        `DEFAULT=${isDefault ? 'YES' : 'NO'},AUTOSELECT=${isDefault ? 'YES' : 'NO'},` +
        `CHANNELS="${track.channels}",URI="${media(track.uri)}"`
    );
  }

  const audioAttribute = manifest.audio.length > 0 ? `,AUDIO="${AudioGroupId}"` : '';

  for (const rendition of manifest.video.renditions) {
    // The rung's own size where it states one, since a player sizes its buffer from what it is
    // told. A film with a single rung states none, and that rung is the source's own size.
    const width = rendition.width && rendition.height ? rendition.width : manifest.video.width;
    const height = rendition.width && rendition.height ? rendition.height : manifest.video.height;

    lines.push(
      `#EXT-X-STREAM-INF:BANDWIDTH=${Math.round(rendition.bitrateKbps * 1000)},` +
        `RESOLUTION=${width}x${height}${audioAttribute}`
    );
    lines.push(media(rendition.uri));
  }

  return { playlist: `${lines.join('\n')}\n`, audioNames };
}

export function masterPlaylistUrl(master: MasterPlaylist): string {
  const blob = new Blob([master.playlist], { type: 'application/vnd.apple.mpegurl' });
  return URL.createObjectURL(blob);
}

File: src/MovieBot.Core/MasterPlaylist.cs
using System.Globalization;
using System.Text;

namespace TheKrystalShip.MovieBot.Core;

/// <summary>
/// Composes the HLS master playlist that binds a title's audio renditions to its video.
///
/// Without one, a player handed the video media playlist alone plays the film silently and has
/// no audio track to switch to — the media playlists carry no reference to each other. The
/// master is what makes the output self-contained: it plays in VLC, in Safari and through
/// hls.js from the same file, rather than requiring every client to compose its own.
///
/// Subtitles stay out of it. They are standalone WebVTT files rather than segmented renditions,
/// and players attach them as text tracks from the manifest.
/// </summary>
public static class MasterPlaylist
{
    public const string FileName = "master.m3u8";

    /// <summary>
    /// Builds the playlist from a manifest. URIs are relative, so the file resolves the same way
    /// whether it is opened from disk or served from <c>/media/{id}/</c>.
    /// </summary>
    public static string Build(Manifest manifest)
    {
        const string audioGroup = "audio";

        var builder = new StringBuilder();
        builder.AppendLine("#EXTM3U");
        builder.AppendLine("#EXT-X-VERSION:7");
        builder.AppendLine("#EXT-X-INDEPENDENT-SEGMENTS");

        var defaultAudio = manifest.Audio.FirstOrDefault(a => a.Default) ?? manifest.Audio.FirstOrDefault();
        var usedNames = new HashSet<string>(StringComparer.Ordinal);

        foreach (var track in manifest.Audio)
        {
            // NAME identifies a rendition within its group, so a repeated label has to be broken
            // apart or a player cannot tell two tracks apart.
            var name = Escape(track.Label);
            while (!usedNames.Add(name))
                name = $"{name} ({track.Id})";

            var isDefault = ReferenceEquals(track, defaultAudio);
            builder.Append("#EXT-X-MEDIA:TYPE=AUDIO")
                .Append(",GROUP-ID=\"").Append(audioGroup).Append('"')
                .Append(",NAME=\"").Append(name).Append('"')
                .Append(",LANGUAGE=\"").Append(Escape(track.Language)).Append('"')
                .Append(",DEFAULT=").Append(isDefault ? "YES" : "NO")
                .Append(",AUTOSELECT=").Append(isDefault ? "YES" : "NO")
                .Append(",CHANNELS=\"").Append(track.Channels.ToString(CultureInfo.InvariantCulture)).Append('"')
                .Append(",URI=\"").Append(track.Uri).Append('"')
                .AppendLine();
        }

        var audioAttribute = manifest.Audio.Count > 0 ? $",AUDIO=\"{audioGroup}\"" : string.Empty;

        foreach (var rendition in manifest.Video.Renditions)
        {
            // CODECS is deliberately absent: a player reads the real codecs from the fMP4
            // initialisation segments, and a declared string that disagrees with them costs a
            // source buffer for no benefit.
            var (width, height) = rendition.SizeIn(manifest.Video);

            builder.Append("#EXT-X-STREAM-INF:BANDWIDTH=")
                .Append((rendition.BitrateKbps * 1000).ToString(CultureInfo.InvariantCulture))
                .Append(",RESOLUTION=")
                .Append(width.ToString(CultureInfo.InvariantCulture))
                .Append('x')
                .Append(height.ToString(CultureInfo.InvariantCulture))
                .Append(audioAttribute)
                .AppendLine();

            builder.AppendLine(rendition.Uri);
        }

        return builder.ToString();
    }

    public static void Write(string outputDirectory, Manifest manifest) =>
        File.WriteAllText(Path.Combine(outputDirectory, FileName), Build(manifest));

    /// <summary>
    /// Attribute values are double-quoted, so an embedded double quote would end the value early.
    /// Real titles carry them: a commentary track named with a quoted nickname is enough.
    /// </summary>
    private static string Escape(string value) => value.Replace('"', '\'');
}
